File: apps/savanaflow/backend/app/schemas/loyalty.py
"""Pydantic v2 schemas for Advanced Loyalty Program — SavanaFlow."""
from __future__ import annotations
from datetime import datetime, date
from typing import Any
from pydantic import BaseModel, EmailStr, Field


# ── Loyalty Program ───────────────────────────────────────────────
class LoyaltyProgramCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    store_id: int | None = None  # Null = all stores
    is_active: bool = True
    start_date: date | None = None
    end_date: date | None = None
    points_per_currency: float = Field(default=0.01, ge=0)
    currency_per_point: float = Field(default=1, ge=0)
    welcome_bonus: int = Field(default=0, ge=0)
    birthday_bonus: int = Field(default=0, ge=0)
    referral_bonus: int = Field(default=0, ge=0)
    points_expiry_days: int = Field(default=0, ge=0)
    max_redemption_percent: int = Field(default=50, ge=0, le=100)


class LoyaltyProgramUpdate(BaseModel):
    name: str | None = Field(None, max_length=200)
    description: str | None = None
    is_active: bool | None = None
    start_date: date | None = None
    end_date: date | None = None
    points_per_currency: float | None = Field(None, ge=0)
    currency_per_point: float | None = Field(None, ge=0)
    welcome_bonus: int | None = Field(None, ge=0)
    birthday_bonus: int | None = Field(None, ge=0)
    referral_bonus: int | None = Field(None, ge=0)
    points_expiry_days: int | None = Field(None, ge=0)
    max_redemption_percent: int | None = Field(None, ge=0, le=100)


class LoyaltyProgramOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    store_id: int | None
    name: str
    description: str | None
    is_active: bool
    start_date: date | None
    end_date: date | None
    points_per_currency: float
    currency_per_point: float
    welcome_bonus: int
    birthday_bonus: int
    referral_bonus: int
    points_expiry_days: int
    max_redemption_percent: int
    created_at: datetime
    tiers: list["LoyaltyTierOut"] = []


# ── Loyalty Tier ───────────────────────────────────────────────────
class LoyaltyTierCreate(BaseModel):
    name: str = Field(min_length=1, max_length=50)
    min_points: int = Field(default=0, ge=0)
    discount_percent: float = Field(default=0, ge=0, le=100)
    points_multiplier: float = Field(default=1, ge=1)
    welcome_bonus: int = Field(default=0, ge=0)
    birthday_bonus: int = Field(default=0, ge=0)
    color: str = Field(default="#6B7280", max_length=10)
    icon: str | None = Field(None, max_length=50)
    order: int = Field(default=0, ge=0)


class LoyaltyTierUpdate(BaseModel):
    name: str | None = Field(None, max_length=50)
    min_points: int | None = Field(None, ge=0)
    discount_percent: float | None = Field(None, ge=0, le=100)
    points_multiplier: float | None = Field(None, ge=1)
    welcome_bonus: int | None = Field(None, ge=0)
    birthday_bonus: int | None = Field(None, ge=0)
    color: str | None = Field(None, max_length=10)
    icon: str | None = None
    order: int | None = Field(None, ge=0)


class LoyaltyTierOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    program_id: int
    name: str
    min_points: int
    discount_percent: float
    points_multiplier: float
    welcome_bonus: int
    birthday_bonus: int
    color: str
    icon: str | None
    order: int


# ── Loyalty Reward ─────────────────────────────────────────────────
class LoyaltyRewardCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    image_url: str | None = Field(None, max_length=500)
    reward_type: str = Field(pattern="^(discount|free_product|voucher|experience)$")
    points_cost: int = Field(gt=0)
    discount_value: float | None = Field(None, ge=0)
    discount_percent: float | None = Field(None, ge=0, le=100)
    product_id: int | None = None
    is_active: bool = True
    valid_for_days: int = Field(default=30, ge=1)
    max_redemptions: int | None = Field(None, gt=0)
    min_tier_id: int | None = None


class LoyaltyRewardUpdate(BaseModel):
    name: str | None = Field(None, max_length=200)
    description: str | None = None
    image_url: str | None = None
    points_cost: int | None = Field(None, gt=0)
    discount_value: float | None = Field(None, ge=0)
    discount_percent: float | None = Field(None, ge=0, le=100)
    product_id: int | None = None
    is_active: bool | None = None
    valid_for_days: int | None = Field(None, ge=1)
    max_redemptions: int | None = Field(None, gt=0)
    min_tier_id: int | None = None


class LoyaltyRewardOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    program_id: int
    name: str
    description: str | None
    image_url: str | None
    reward_type: str
    points_cost: int
    discount_value: float | None
    discount_percent: float | None
    product_id: int | None
    is_active: bool
    valid_for_days: int
    max_redemptions: int | None
    current_redemptions: int
    min_tier_id: int | None
    created_at: datetime


# ── Loyalty Card ───────────────────────────────────────────────────
class LoyaltyCardCreate(BaseModel):
    program_id: int
    customer_id: int
    referred_by_card_number: str | None = None  # For referral tracking


class LoyaltyCardOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    program_id: int
    customer_id: int
    card_number: str
    current_tier_id: int | None
    points_balance: int
    total_points_earned: int
    total_points_redeemed: int
    total_visits: int
    total_spent: float
    member_since: date
    last_visit: datetime | None
    is_active: bool
    referral_count: int
    created_at: datetime
    # Computed fields
    customer: "POSCustomerBrief | None" = None
    current_tier: "LoyaltyTierOut | None" = None


class POSCustomerBrief(BaseModel):
    """Brief customer info for nested responses."""
    model_config = {"from_attributes": True}
    id: int
    name: str
    phone: str | None
    email: str | None
    loyalty_tier: str


# ── Loyalty Card Transaction ───────────────────────────────────────
class LoyaltyCardTransactionOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    card_id: int
    transaction_type: str
    points: int
    source_type: str | None
    source_id: int | None
    description: str | None
    balance_after: int
    expires_at: datetime | None
    created_at: datetime


# ── Loyalty Redemption ─────────────────────────────────────────────
class LoyaltyRedemptionCreate(BaseModel):
    card_id: int
    reward_id: int


class LoyaltyRedemptionOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    card_id: int
    reward_id: int
    sale_id: int | None
    points_used: int
    voucher_code: str | None
    status: str
    expires_at: datetime | None
    used_at: datetime | None
    created_at: datetime
    reward: "LoyaltyRewardOut | None" = None


# ── POS Integration ────────────────────────────────────────────────
class SaleEarnPoints(BaseModel):
    """Request to earn points from a sale."""
    card_number: str
    sale_id: int
    sale_amount: float = Field(gt=0)


class SaleRedeemPoints(BaseModel):
    """Request to redeem points in a sale."""
    card_number: str
    points_to_use: int = Field(gt=0)
    sale_amount: float = Field(gt=0)


class LoyaltySaleResult(BaseModel):
    """Result of loyalty operation in a sale."""
    points_earned: int
    points_used: int
    loyalty_discount: float
    new_balance: int
    tier_name: str | None
    tier_multiplier: float


class CardLookupResult(BaseModel):
    """Result of card lookup for POS."""
    card_number: str
    customer_name: str
    customer_phone: str | None
    points_balance: int
    tier_name: str
    tier_color: str
    tier_discount_percent: float
    points_value: float  # Currency value of available points
    can_redeem: bool


# ── Stats ──────────────────────────────────────────────────────────
class LoyaltyStats(BaseModel):
    """Loyalty program statistics."""
    total_members: int
    active_members: int
    total_points_earned: int
    total_points_redeemed: int
    total_points_balance: int
    total_redemptions: int
    tier_distribution: list[dict[str, Any]]
    top_members: list[dict[str, Any]]
    recent_transactions: list[dict[str, Any]]
    rewards_stats: dict[str, Any]


# ── Referral ───────────────────────────────────────────────────────
class ReferralCreate(BaseModel):
    referrer_card_number: str
    referred_customer_id: int


class ReferralOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    referrer_card_id: int
    referred_customer_id: int
    status: str
    bonus_awarded: bool
    bonus_points: int
    first_purchase_amount: float | None
    first_purchase_at: datetime | None
    created_at: datetime


# ── Update forward references ──────────────────────────────────────
LoyaltyProgramOut.model_rebuild()
LoyaltyCardOut.model_rebuild()
LoyaltyRedemptionOut.model_rebuild()

File: apps/savanaflow/frontend/src/store/loyalty.ts
import { create } from 'zustand'
import api from '../lib/api'
import toast from 'react-hot-toast'

// Types
export interface LoyaltyProgram {
  id: number
  store_id: number | null
  name: string
  description: string | null
  is_active: boolean
  start_date: string | null
  end_date: string | null
  points_per_currency: number
  currency_per_point: number
  welcome_bonus: number
  birthday_bonus: number
  referral_bonus: number
  points_expiry_days: number
  max_redemption_percent: number
  created_at: string
  tiers: LoyaltyTier[]
}

export interface LoyaltyTier {
  id: number
  program_id: number
  name: string
  min_points: number
  discount_percent: number
  points_multiplier: number
  welcome_bonus: number
  birthday_bonus: number
  color: string
  icon: string | null
  order: number
}

export interface LoyaltyReward {
  id: number
  program_id: number
  name: string
  description: string | null
  image_url: string | null
  reward_type: 'discount' | 'free_product' | 'voucher' | 'experience'
  points_cost: number
  discount_value: number | null
  discount_percent: number | null
  product_id: number | null
  is_active: boolean
  valid_for_days: number
  max_redemptions: number | null
  current_redemptions: number
  min_tier_id: number | null
  created_at: string
}

export interface LoyaltyCard {
  id: number
  program_id: number
  customer_id: number
  card_number: string
  current_tier_id: number | null
  points_balance: number
  total_points_earned: number
  total_points_redeemed: number
  total_visits: number
  total_spent: number
  member_since: string
  last_visit: string | null
  is_active: boolean
  referral_count: number
  created_at: string
  customer?: {
    id: number
    name: string
    phone: string | null
    email: string | null
    loyalty_tier: string
  }
  current_tier?: LoyaltyTier
}

export interface LoyaltyTransaction {
  id: number
  card_id: number
  transaction_type: 'earn' | 'redeem' | 'bonus' | 'expire' | 'adjust'
  points: number
  source_type: string | null
  source_id: number | null
  description: string | null
  balance_after: number
  expires_at: string | null
  created_at: string
}

export interface CardLookupResult {
  card_number: string
  customer_name: string
  customer_phone: string | null
  points_balance: number
  tier_name: string
  tier_color: string
  tier_discount_percent: number
  points_value: number
  can_redeem: boolean
}

export interface LoyaltyStats {
  total_members: number
  active_members: number
  total_points_earned: number
  total_points_redeemed: number
  total_points_balance: number
  total_redemptions: number
  tier_distribution: Array<{ tier: string; color: string; count: number }>
  top_members: Array<{
    card_id: number
    customer_name: string
    points_balance: number
    total_spent: number
    tier: string
  }>
  recent_transactions: Array<{
    id: number
    card_number: string | null
    customer_name: string | null
    type: string
    points: number
    created_at: string
  }>
  rewards_stats: {
    total: number
    active: number
  }
}

interface LoyaltyState {
  // Programs
  programs: LoyaltyProgram[]
  currentProgram: LoyaltyProgram | null
  programsLoading: boolean

  // Tiers
  tiers: LoyaltyTier[]

  // Rewards
  rewards: LoyaltyReward[]
  rewardsLoading: boolean

  // Cards
  cards: LoyaltyCard[]
  cardsTotal: number
  cardsLoading: boolean
  currentCard: LoyaltyCard | null
  cardTransactions: LoyaltyTransaction[]

  // Stats
  stats: LoyaltyStats | null
  statsLoading: boolean

  // POS Integration
  scannedCard: CardLookupResult | null
  scannedCardLoading: boolean

  // Actions
  fetchPrograms: (storeId?: number) => Promise<void>
  fetchProgram: (programId: number) => Promise<void>
  createProgram: (data: Partial<LoyaltyProgram>) => Promise<LoyaltyProgram | null>
  updateProgram: (programId: number, data: Partial<LoyaltyProgram>) => Promise<LoyaltyProgram | null>
  deleteProgram: (programId: number) => Promise<boolean>

  // Tiers
  createTier: (programId: number, data: Partial<LoyaltyTier>) => Promise<LoyaltyTier | null>
  updateTier: (tierId: number, data: Partial<LoyaltyTier>) => Promise<LoyaltyTier | null>
  deleteTier: (tierId: number) => Promise<boolean>

  // Rewards
  fetchRewards: (programId?: number) => Promise<void>
  createReward: (programId: number, data: Partial<LoyaltyReward>) => Promise<LoyaltyReward | null>
  updateReward: (rewardId: number, data: Partial<LoyaltyReward>) => Promise<LoyaltyReward | null>
  deleteReward: (rewardId: number) => Promise<boolean>

  // Cards
  fetchCards: (params?: { programId?: number; tierId?: number; search?: string; page?: number; size?: number }) => Promise<void>
  fetchCard: (cardId: number) => Promise<void>
  registerCard: (data: { program_id: number; customer_id: number; referred_by_card_number?: string }) => Promise<LoyaltyCard | null>
  adjustPoints: (cardId: number, points: number, description: string) => Promise<boolean>
  fetchCardTransactions: (cardId: number, page?: number) => Promise<void>

  // Stats
  fetchStats: (programId?: number) => Promise<void>

  // POS Integration
  lookupCard: (cardNumber: string) => Promise<CardLookupResult | null>
  earnPoints: (cardNumber: string, saleId: number, saleAmount: number) => Promise<{ points_earned: number; new_balance: number } | null>
  redeemPoints: (cardNumber: string, pointsToUse: number, saleAmount: number) => Promise<{ points_used: number; loyalty_discount: number; new_balance: number } | null>
  clearScannedCard: () => void

  // Reset
  reset: () => void
}

const initialState = {
  programs: [],
  currentProgram: null,
  programsLoading: false,
  tiers: [],
  rewards: [],
  rewardsLoading: false,
  cards: [],
  cardsTotal: 0,
  cardsLoading: false,
  currentCard: null,
  cardTransactions: [],
  stats: null,
  statsLoading: false,
  scannedCard: null,
  scannedCardLoading: false,
}

export const useLoyaltyStore = create<LoyaltyState>((set, get) => ({
  ...initialState,

  // Programs
  fetchPrograms: async (storeId?: number) => {
    set({ programsLoading: true })
    try {
      const params = storeId ? { store_id: storeId } : {}
      const { data } = await api.get('/loyalty/programs', { params })
      set({ programs: data, programsLoading: false })
    } catch (error) {
      set({ programsLoading: false })
      toast.error('Erreur lors du chargement des programmes')
    }
  },

  fetchProgram: async (programId: number) => {
    set({ programsLoading: true })
    try {
      const { data } = await api.get(`/loyalty/programs/${programId}`)
      set({ currentProgram: data, tiers: data.tiers || [], programsLoading: false })
    } catch (error) {
      set({ programsLoading: false })
      toast.error('Erreur lors du chargement du programme')
    }
  },

  createProgram: async (data: Partial<LoyaltyProgram>) => {
    try {
      const { data: program } = await api.post('/loyalty/programs', data)
      set(state => ({ programs: [...state.programs, program] }))
      toast.success('Programme créé avec succès')
      return program
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de la création')
      return null
    }
  },

  updateProgram: async (programId: number, data: Partial<LoyaltyProgram>) => {
    try {
      const { data: program } = await api.put(`/loyalty/programs/${programId}`, data)
      set(state => ({
        programs: state.programs.map(p => p.id === programId ? program : p),
        currentProgram: state.currentProgram?.id === programId ? program : state.currentProgram,
      }))
      toast.success('Programme mis à jour')
      return program
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de la mise à jour')
      return null
    }
  },

  deleteProgram: async (programId: number) => {
    try {
      await api.delete(`/loyalty/programs/${programId}`)
      set(state => ({
        programs: state.programs.filter(p => p.id !== programId),
        currentProgram: state.currentProgram?.id === programId ? null : state.currentProgram,
      }))
      toast.success('Programme supprimé')
      return true
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de la suppression')
      return false
    }
  },

  // Tiers
  createTier: async (programId: number, data: Partial<LoyaltyTier>) => {
    try {
      const { data: tier } = await api.post(`/loyalty/programs/${programId}/tiers`, data)
      set(state => ({
        tiers: [...state.tiers, tier].sort((a, b) => a.order - b.order),
        currentProgram: state.currentProgram
          ? { ...state.currentProgram, tiers: [...(state.currentProgram.tiers || []), tier] }
          : null,
      }))
      toast.success('Niveau créé avec succès')
      return tier
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de la création')
      return null
    }
  },

  updateTier: async (tierId: number, data: Partial<LoyaltyTier>) => {
    try {
      const { data: tier } = await api.put(`/loyalty/tiers/${tierId}`, data)
      set(state => ({
        tiers: state.tiers.map(t => t.id === tierId ? tier : t),
      }))
      toast.success('Niveau mis à jour')
      return tier
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de la mise à jour')
      return null
    }
  },

  deleteTier: async (tierId: number) => {
    try {
      await api.delete(`/loyalty/tiers/${tierId}`)
      set(state => ({
        tiers: state.tiers.filter(t => t.id !== tierId),
      }))
      toast.success('Niveau supprimé')
      return true
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de la suppression')
      return false
    }
  },

  // Rewards
  fetchRewards: async (programId?: number) => {
    set({ rewardsLoading: true })
    try {
      const params = programId ? { program_id: programId } : {}
      const { data } = await api.get('/loyalty/rewards', { params })
      set({ rewards: data, rewardsLoading: false })
    } catch (error) {
      set({ rewardsLoading: false })
      toast.error('Erreur lors du chargement des récompenses')
    }
  },

  createReward: async (programId: number, data: Partial<LoyaltyReward>) => {
    try {
      const { data: reward } = await api.post(`/loyalty/programs/${programId}/rewards`, data)
      set(state => ({ rewards: [...state.rewards, reward] }))
      toast.success('Récompense créée avec succès')
      return reward
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de la création')
      return null
    }
  },

  updateReward: async (rewardId: number, data: Partial<LoyaltyReward>) => {
    try {
      const { data: reward } = await api.put(`/loyalty/rewards/${rewardId}`, data)
      set(state => ({ rewards: state.rewards.map(r => r.id === rewardId ? reward : r) }))
      toast.success('Récompense mise à jour')
      return reward
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de la mise à jour')
      return null
    }
  },

  deleteReward: async (rewardId: number) => {
    try {
      await api.delete(`/loyalty/rewards/${rewardId}`)
      set(state => ({ rewards: state.rewards.filter(r => r.id !== rewardId) }))
      toast.success('Récompense supprimée')
      return true
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de la suppression')
      return false
    }
  },

  // Cards
  fetchCards: async (params?: { programId?: number; tierId?: number; search?: string; page?: number; size?: number }) => {
    set({ cardsLoading: true })
    try {
      const { data } = await api.get('/loyalty/cards', { params })
      set({ cards: data.items, cardsTotal: data.total, cardsLoading: false })
    } catch (error) {
      set({ cardsLoading: false })
      toast.error('Erreur lors du chargement des cartes')
    }
  },

  fetchCard: async (cardId: number) => {
    set({ cardsLoading: true })
    try {
      const { data } = await api.get(`/loyalty/cards/${cardId}`)
      set({ currentCard: data, cardsLoading: false })
    } catch (error) {
      set({ cardsLoading: false })
      toast.error('Erreur lors du chargement de la carte')
    }
  },

  registerCard: async (data: { program_id: number; customer_id: number; referred_by_card_number?: string }) => {
    try {
      const { data: card } = await api.post('/loyalty/cards/register', data)
      set(state => ({ cards: [card, ...state.cards], cardsTotal: state.cardsTotal + 1 }))
      toast.success('Client inscrit au programme')
      return card
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de l\'inscription')
      return null
    }
  },

  adjustPoints: async (cardId: number, points: number, description: string) => {
    try {
      const { data: card } = await api.post(`/loyalty/cards/${cardId}/adjust`, null, {
        params: { points, description },
      })
      set(state => ({
        cards: state.cards.map(c => c.id === cardId ? card : c),
        currentCard: state.currentCard?.id === cardId ? card : state.currentCard,
      }))
      toast.success('Points ajustés avec succès')
      return true
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de l\'ajustement')
      return false
    }
  },

  fetchCardTransactions: async (cardId: number, page: number = 1) => {
    try {
      const { data } = await api.get(`/loyalty/cards/${cardId}/transactions`, {
        params: { page, size: 50 },
      })
      set({ cardTransactions: data.items })
    } catch (error) {
      toast.error('Erreur lors du chargement des transactions')
    }
  },

  // Stats
  fetchStats: async (programId?: number) => {
    set({ statsLoading: true })
    try {
      const params = programId ? { program_id: programId } : {}
      const { data } = await api.get('/loyalty/stats', { params })
      set({ stats: data, statsLoading: false })
    } catch (error) {
      set({ statsLoading: false })
      toast.error('Erreur lors du chargement des statistiques')
    }
  },

  // POS Integration
  lookupCard: async (cardNumber: string) => {
    set({ scannedCardLoading: true })
    try {
      const { data } = await api.get(`/loyalty/cards/lookup/${cardNumber}`)
      set({ scannedCard: data, scannedCardLoading: false })
      return data
    } catch (error: any) {
      set({ scannedCardLoading: false, scannedCard: null })
      toast.error(error.response?.data?.detail || 'Carte non trouvée')
      return null
    }
  },

  earnPoints: async (cardNumber: string, saleId: number, saleAmount: number) => {
    try {
      const { data } = await api.post('/loyalty/sale/earn', {
        card_number: cardNumber,
        sale_id: saleId,
        sale_amount: saleAmount,
      })
      // Update scanned card balance
      set(state => ({
        scannedCard: state.scannedCard
          ? { ...state.scannedCard, points_balance: data.new_balance }
          : null,
      }))
      return { points_earned: data.points_earned, new_balance: data.new_balance }
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de l\'attribution des points')
      return null
    }
  },

  redeemPoints: async (cardNumber: string, pointsToUse: number, saleAmount: number) => {
    try {
      const { data } = await api.post('/loyalty/sale/redeem', {
        card_number: cardNumber,
        points_to_use: pointsToUse,
        sale_amount: saleAmount,
      })
      // Update scanned card balance
      set(state => ({
        scannedCard: state.scannedCard
          ? { ...state.scannedCard, points_balance: data.new_balance }
          : null,
      }))
      return { points_used: data.points_used, loyalty_discount: data.loyalty_discount, new_balance: data.new_balance }
    } catch (error: any) {
      toast.error(error.response?.data?.detail || 'Erreur lors de l\'utilisation des points')
      return null
    }
  },

  clearScannedCard: () => set({ scannedCard: null }),

  reset: () => set(initialState),
}))
